### DISCUSSION TITLE
Reason for separate `tools/<port>` directory?

### DISCUSSION THREAD

**Discussion by ekilmer** [microsoft/vcpkg#24448] [Q&A]
Hello, I'm trying to understand why vcpkg has a separate directory for executables/tools, i.e. the `tools/<port>` directory, as opposed to throwing tools into the `bin` directory, as is customary on Unix OSs. To be fair, I've thought only a little bit about different scenarios, so maybe there's an obvious answer (possibly already existing somewhere), in which case, thank you for posting it here for me.  

I understand vcpkg is meant to be used primarily as a way to distribute _libraries_ and not tools, but there are a few libraries that require a runnable tool to function well, like `protoc` with the protobuf project.

By having a separate `tools/<port>` directory for tools and `bin` directory for DLLs (on Windows) it goes against the common CMake convention of using `GNUInstallDirs`'s defined `CMAKE_INSTALL_BINDIR` value, which is for "user executables" and CMake's `install` command for `RUNTIME` target "Executables" and "DLLS (on all Windows-based systems [...])", i.e. there is no way to separately specify where to install DLLs and executables.

Accordingly, vcpkg must move the tools and fix the CMake Config file paths in order for the generated export CMake files to function correctly. This doesn't sound too worrisome because projects are only interfacing with CMake via targets and the paths don't matter after they're already fixed, but there _is_ an issue if a tool expects a resource file to be located relative to its current location, which breaks when changing from a single parent directory (`bin`) to two parent directories (`tools/<port>`). Sure, `GNUInstallDirs` is optional and not required/always included, but it seems to be the only widely-used module of its kind, and the convention provides consistent and standardized variable names for all CMake projects to use and provide as configurable options to users.

Moreover, projects requiring relative paths from the executable for resource-file lookups should ideally have configuration options to specify the installation location of the tools (LLVM does this with `LLVM_TOOLS_INSTALL_DIR`) and perform relative path calculations at configure-time, but that is quite advanced and not at the top of developers minds considering `GNUInstallDirs` defaults to tools _and_ DLLs living at `CMAKE_INSTALL_BINDIR`, which [vcpkg sets as `bin`](https://github.com/microsoft/vcpkg/blob/fb02a785618a1fbaced32b206171e09fd57d4622/scripts/cmake/vcpkg_configure_cmake.cmake#L316). Maybe vcpkg should set `CMAKE_INSTALL_BINDIR` to `tools/<port>` by default?

Another concern is with Windows. The same DLLs must exist in at least two places, the `bin` directory and the `tools/<port>` directory, next to the tool itself for a package. This also might not seem like a big deal, but when a tool depends on other libraries, that other library's DLLs must be copied and duplicated to live next to the tool's unique `tools/<port>` directory. This duplication issue could be solved by just leaving the tool(s) in `bin` next to the DLLs that are already installed.

I've tried searching for discussion about this previously, but the only one I could find was https://github.com/microsoft/vcpkg/issues/59#issuecomment-248720356 and described the reason for a separate tools directory as "upgrading the DLLs is tricky", and maybe this is true, but I don't know the specifics of what is actually "tricky" about it---could someone explain this more, please?

**Comment by Osyotr** [comment]
Related: https://github.com/microsoft/vcpkg/issues/17607

**Comment by BillyONeal** [comment]
This is before my time so take with a grain of salt, but I believe this is the rationale:

* `vcpkg` is not an *application* deployment system, and we intentionally want to discourage people from using it that way because we would get asked to make changes in support of that experience that would damage the primary use case we care about of deploying libraries.
  * For example, we rejected an attempt to add a gcc port because it wasn't meaningful to use as a library, and don't make the clang produced by the LLVM port practically usable.
* We want to avoid cases where ports accidentally depend on each other's output, and so want an explicit request/opt-in to use a tool from a dependency.
* Tools produced by a port aren't intended to be exposed to vcpkg's downstream customers so there's no build system exposed / contractual reason they all have to be in the "bin" directory, unlike DLLs or shared objects or dylibs.
* Our convention for everything else is "root/port-name/whatever goes here"

As @Osyotr points out, https://github.com/microsoft/vcpkg/issues/17607 discusses some current places that creates issues, particularly for ports like qt where a tool and its supporting libraries themselves don't come from the same port.

**Reply by gracicot** [reply]
What about tools required for CMake configuration such as code generators? I have an example where I a cross compile, but need a host binary available to do code generation. Right now trying to do `find_package(code-generator)` fails unless I manually add prefix paths.

**Reply by Neumann-A** [reply]
This shouldnt fail if done correctly. (e.g. see Qt6)

**Reply by gracicot** [reply]
As far as I can see, host dependencies cannot be discovered using `find_package`

**Reply by Neumann-A** [reply]
It can you have to use `find_package` just with the right arguments (e.g. `PATHS`) and hint it into the right direction. Qt for example has `QT_HOST_PATH`. Your CMake script needs to be aware that there needs to be basically two different `find_package` calls. One for the target one for the host. While the target search always works out of the box the host search needs extra setup from your side.

**Reply by gracicot** [reply]
I see. Thanks for the tip. I'll try to see what Qt ports are doing.
